File: cadwiki-nuget/cadwiki.DllReloader/AutoCAD/UiRibbon/Buttons/GenericClickCommandHandler.vb
'Option Strict On
'Option Infer Off
'Option Explicit On

Imports System.IO
Imports System.Reflection
Imports System.Windows.Input
Imports Autodesk.AutoCAD.ApplicationServices
Imports Autodesk.Windows

Namespace AutoCAD.UiRibbon.Buttons
    Public Class GenericClickCommandHandler
        Implements ICommand

        Public Sub New()
        End Sub

        Public Function CanExecute(parameter As Object) As Boolean Implements ICommand.CanExecute
            Return True
        End Function

        Public Event CanExecuteChanged(sender As Object, e As EventArgs) Implements ICommand.CanExecuteChanged


        Public Sub Execute(parameter As Object) Implements ICommand.Execute
            Dim uiRouter As UiRouter = Nothing
            Dim netReloader As AutoCADAppDomainDllReloader = Nothing
            Try
                If TypeOf parameter Is RibbonButton Then
                    Dim button As RibbonButton = TryCast(parameter, RibbonButton)
                    netReloader.Log("GenericClickCommandHandler Executing Method.")
                    uiRouter = button.CommandParameter
                    netReloader = uiRouter.NetReloader
                    Dim assemblyName As String = uiRouter.AssemblyName

                    If (uiRouter.FullClassName Is Nothing) Then
                        netReloader.Log("Full class name was null. Make sure you're Ribbon button full class name is correct.")
                        Return
                    End If
                    netReloader.Log("Full class name: " & uiRouter.FullClassName)

                    If (uiRouter.MethodName Is Nothing) Then
                        netReloader.Log("MethodName was null. Make sure you're Ribbon button full class name is correct.")
                        Return
                    End If
                    netReloader.Log("Method name: " & uiRouter.MethodName)

                    Dim dllRepo As String = Path.GetDirectoryName(netReloader.GetDllPath())
                    Dim asm As Assembly = AcadAssemblyUtils.GetNewestAssembly(AppDomain.CurrentDomain.GetAssemblies(), assemblyName,
                                                            dllRepo + "\" + assemblyName + ".dll")
                    'Dim asm As System.Reflection.Assembly = If(App.ReloadedAssembly, Assembly.GetExecutingAssembly)
                    Dim types As Type() = cadwiki.NetUtils.AssemblyUtils.GetTypesSafely(asm)
                    Dim type As Type = asm.GetType(uiRouter.FullClassName)
                    Dim methodInfo As MethodInfo = type.GetMethod(uiRouter.MethodName)
                    If methodInfo = Nothing Then
                        netReloader.Log("Method not found: " & uiRouter.MethodName)
                    Else
                        Dim o As Object = Activator.CreateInstance(type)
                        If (uiRouter.Parameters IsNot Nothing) Then
                            methodInfo.Invoke(o, uiRouter.Parameters)
                        Else
                            methodInfo.Invoke(o, Nothing)
                        End If
                    End If
                End If
            Catch ex As Exception
                Dim window As cadwiki.WpfUi.Templates.WindowAutoCADException =
                        New WpfUi.Templates.WindowAutoCADException(ex)
                window.Show()
                netReloader.Log("Exception: " & ex.Message)
                If (ex.Message.Equals("The path is not of a legal form.")) Then
                    netReloader.Log("Mostly likely caused by incorrect method name in UiRouter object.")
                    netReloader.Log("Double check that the Method name and Full class name above are correct.")
                Else
                    netReloader.Log("Mostly likely caused by incorrect solution name in UiRouter object: " &
                            netReloader.GetIExtensionApplicationClassName())
                End If

                If (uiRouter IsNot Nothing) Then
                    netReloader.Log("UiRouter object: " & uiRouter.FullClassName)
                End If
            End Try

        End Sub



        Sub CallMethod(f As Action)
            Dim doc As Document = Application.DocumentManager.MdiActiveDocument
            doc.Editor.WriteMessage(vbLf & "Calling...")
            f()
        End Sub


    End Class
End Namespace

File: cadwiki-nuget/cadwiki.DllReloader/AutoCAD/UiRibbon/Buttons/GenericClickCommandHandler.cs
using System;
// Option Strict On
// Option Infer Off
// Option Explicit On

using System.IO;
using System.Windows.Input;
using Autodesk.Windows;
using Microsoft.VisualBasic;

namespace cadwiki.DllReloader.AutoCAD.UiRibbon.Buttons
{
    public class GenericClickCommandHandler : ICommand
    {

        public GenericClickCommandHandler()
        {
        }

        public bool CanExecute(object parameter)
        {
            return true;
        }

        public event EventHandler CanExecuteChanged;


        public void Execute(object parameter)
        {
            UiRouter uiRouter = null;
            var netReloader = new AutoCADAppDomainDllReloader();
            netReloader.Log("GenericClickCommandHandler Executing Method.");
            try
            {
                if (parameter is null)
                {
                    netReloader.Log("Parameter was null.");
                    return;
                }

                if (parameter is RibbonButton)
                {

                    RibbonButton button = parameter as RibbonButton;
                    if (button is null)
                    {
                        netReloader.Log("Button was null.");
                        return;
                    }

                    uiRouter = (UiRouter)button.CommandParameter;
                    netReloader = uiRouter.NetReloader;

                    string assemblyName = uiRouter.AssemblyName;

                    if (uiRouter.FullClassName is null)
                    {
                        netReloader.Log("Full class name was null. Make sure you're Ribbon button full class name is correct.");
                        return;
                    }
                    netReloader.Log("Full class name: " + uiRouter.FullClassName);

                    if (uiRouter.MethodName is null)
                    {
                        netReloader.Log("MethodName was null. Make sure you're Ribbon button full class name is correct.");
                        return;
                    }
                    netReloader.Log("Method name: " + uiRouter.MethodName);

                    string dllRepo = Path.GetDirectoryName(netReloader.GetDllPath());
                    var asm = AcadAssemblyUtils.GetNewestAssembly(AppDomain.CurrentDomain.GetAssemblies(), assemblyName, dllRepo + @"\" + assemblyName + ".dll");
                    if (asm == null)
                    {
                        throw new Exception(string.Format("Assembly {0} not found in folder {1} ", 
                            assemblyName, dllRepo));
                    }
                    Type[] types = NetUtils.AssemblyUtils.GetTypesSafely(asm);
                    if (types == null)
                    {
                        throw new Exception(string.Format("Unable to parse types from Assembly {0} in folder {1} ", 
                            assemblyName, dllRepo));
                    }
                    var @type = asm.GetType(uiRouter.FullClassName);
                    if (@type == null)
                    {
                        throw new Exception(string.Format("Unable to parse class {0} from Assembly {1} in folder {2} ", 
                            uiRouter.FullClassName, assemblyName, dllRepo));
                    }
                    var methodInfo = type.GetMethod(uiRouter.MethodName);
                    if (methodInfo == null)
                    {
                        netReloader.Log("Method not found: " + uiRouter.MethodName);
                        throw new Exception(string.Format("Unable to parse method {0} from class {1} from Assembly {2} in folder {3} ", 
                            uiRouter.MethodName, uiRouter.FullClassName, assemblyName, dllRepo));
                    }
                    else
                    {
                        var o = Activator.CreateInstance(type);
                        if (uiRouter.Parameters is not null)
                        {
                            methodInfo.Invoke(o, uiRouter.Parameters);
                        }
                        else
                        {
                            methodInfo.Invoke(o, null);
                        }
                    }
                }
            }
            catch (Exception ex)
            {
                var window = new WpfUi.Templates.WindowAutoCADException(ex);
                window.Show();

                if (ex.Message is not null)
                {
                    netReloader.Log("Exception: " + ex.Message);
                    if (ex.Message.Equals("The path is not of a legal form."))
                    {
                        netReloader.Log("Mostly likely caused by incorrect method name in UiRouter object.");
                        netReloader.Log("Double check that the Method name and Full class name above are correct.");
                    }
                    else
                    {
                        netReloader.Log("Mostly likely caused by incorrect solution name in UiRouter object: " + netReloader.GetIExtensionApplicationClassName());
                    }
                }
                else
                {

                }


                if (uiRouter is not null)
                {
                    netReloader.Log("UiRouter object: " + uiRouter.FullClassName);
                }
            }

        }



        public void CallMethod(Action f)
        {
            var doc = Autodesk.AutoCAD.ApplicationServices.Core.Application.DocumentManager.MdiActiveDocument;
            doc.Editor.WriteMessage(Environment.NewLine + "Calling...");
            f();
        }


    }
}